### PR TITLE
Adds in-game bug reports without needing a GitHub account

### PR DESCRIPTION
**Pull request by vero5123**
# About the pull request

title

# Explain why it's good for the game
 Essentially, the idea is to allow for the creation of bug reports in-game without needing to own a Github account, this can help to incentivize issue reporting and be beneficial overall. 

TODO:
- [x] Implement the Github API token to allow for anonymous bug reports
- [x] Make in-game bug reports admin approval only to prevent abuse and to make sure only quality bug reports are approved.

Notable changes from Goonstation's bug report system is that I migrated it to type script and it now requires admin approval.
# Changelog
Although I made alterations to meet our needs, all credit goes to Pali for their amazing work and to the devs at Goonstation. 
:cl:
add: Adds the ability for users to make bug reports in-game
ui: New bug report system ui
/:cl:

### CODE REVIEW

**Comment by harryob** [comment]
this requires the original devs of this feature on goon to license this agpl-ey, goon's cc-by-nc-sa is not compatible

**Comment by pali6** [comment]
Heya, original author of the goonstation feature here. You have my permission to port this under whichever license you use. Enjoy.

**Comment by vero5123** [comment]
> Heya, original author of the goonstation feature here. You have my permission to port this under whichever license you use. Enjoy.

Thank you!

**Comment by vero5123** [comment]
How to test this pr:
Create a personal token - give it access to your forked repo and make sure issues are enabled on the fork.

In the config file add 

# this is your api token
GITHUB_APP_API 
# name of the repo
REPO_NAME cmss13
# the owner of the repo, so if it's a forked repo it would be your username
ORG

See https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#create-an-issue for more details

File: code/datums/bug_report.dm [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	request.prepare(RUSTG_HTTP_METHOD_POST, url, json_encode(payload), headers)
+	request.begin_async()
+	UNTIL(request.is_complete())
+	var/datum/http_response/response = request.into_response()
+
+	if(response.errored || response.status_code != STATUS_SUCCESS)
+		external_link_prompt(user)
+	else
+		message_admins("[user.ckey] has approved a bug report from [initial_user.ckey] titled [bug_report_data["title"]] at [time2text(world.timeofday, "YYYY-MM-DD hh:mm:ss")].")
+		to_chat(initial_user, SPAN_WARNING("An admin has successfully submitted your report and it should now be visible on GitHub, Thanks again!"))
```
Comment (harryob) [inline code/datums/bug_report.dm:115]:
```suggestion
		to_chat(initial_user, SPAN_WARNING("An admin has successfully submitted your report and it should now be visible on GitHub. Thanks again!"))
```

File: code/datums/bug_report.dm [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	qdel(src)// approved and submitted, we no longer need the datum.
+
+// proc that creates a ticket for an admin to approve or deny a bug report request
+/datum/tgui_bug_report_form/proc/bug_report_request()
+	to_chat(initial_user, SPAN_WARNING("Your bug report has been submitted, thank you!"))
+
+	GLOB.bug_reports += src
+	var/general_message = "[initial_user.ckey] has created a bug report, you may modify the report to your liking before submitting it to Github"
```
Comment (harryob) [inline code/datums/bug_report.dm:123]:
```suggestion
	var/general_message = "[initial_user.ckey] has created a bug report, you may modify the report to your liking before submitting it to GitHub."
```

File: code/modules/admin/topic/topic.dm [hunk 3]
```diff
@@ -2248,6 +2248,22 @@
...
+		if(!check_rights(R_ADMIN|R_MOD))
+			return
+
+		var/datum/tgui_bug_report_form/bug_report = locate(href_list["view_bug_report"])
+		if(QDELETED(bug_report))
+			to_chat(usr, SPAN_WARNING("This bug report is no longer available"))
+			return
+
+		if(bug_report.admin_user)
+			to_chat(usr, SPAN_WARNING("Another administrator is currently accessing this report, please wait for them to finish before making any changes"))
```
Comment (harryob) [inline code/modules/admin/topic/topic.dm:2261]:
```suggestion
			to_chat(usr, SPAN_WARNING("Another administrator is currently accessing this report, please wait for them to finish before making any changes."))
```

File: code/modules/admin/topic/topic.dm [hunk 4]
```diff
@@ -2248,6 +2248,22 @@
 			return
 		return remove_tagged_datum(datum_to_remove)
 
+	if(href_list["view_bug_report"])
+		if(!check_rights(R_ADMIN|R_MOD))
+			return
+
+		var/datum/tgui_bug_report_form/bug_report = locate(href_list["view_bug_report"])
+		if(QDELETED(bug_report))
+			to_chat(usr, SPAN_WARNING("This bug report is no longer available"))
```
Comment (harryob) [inline code/modules/admin/topic/topic.dm:2257]:
```suggestion
			to_chat(usr, SPAN_WARNING("This bug report is no longer available."))
```

File: tgui/packages/tgui/interfaces/BugReportForm.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        <Section fill scrollable>
+          <Flex direction="column" height="100%">
+            <Flex.Item className="text-center">
+              <a
+                href="https://github.com/cmss13-devs/cmss13/issues"
+                target="_blank"
+                rel="noreferrer"
+                className="link"
+              >
+                Github Repository
```
Comment (harryob) [inline tgui/packages/tgui/interfaces/BugReportForm.tsx:75]:
```suggestion
                GitHub Repository
```

File: tgui/packages/tgui/interfaces/BugReportForm.tsx [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    report_details?.description || '',
+  );
+  const [expected_behavior, setExpectedBehavior] = useState(
+    report_details?.expected_behavior || '',
+  );
+
+  const titleRef = createRef<HTMLInputElement>();
+  const stepsRef = createRef<HTMLTextAreaElement>();
+  const descriptionRef = createRef<HTMLInputElement>();
+  const expectedBehaviorRef = createRef<HTMLInputElement>();
```
Comment (harryob) [inline tgui/packages/tgui/interfaces/BugReportForm.tsx:47]: where are these refs used

File: code/datums/bug_report.dm [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	initial_user.current_ticket.AddInteraction(href_message)
+
+/datum/tgui_bug_report_form/ui_act(action, list/params, datum/tgui/ui)
+	. = ..()
+	if (.)
+		return
+	var/mob/user = ui.user
+	switch(action)
+		if("confirm")
+			bug_report_data = params
```
Comment (harryob) [inline code/datums/bug_report.dm:135]: can we do some additional sanity checks on this? this is later sent raw to admins, so could be sending intentionally malformed XSS text to admins chats, which is bad

File: code/datums/bug_report.dm [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+// the real deal, we are sending the request through the api.
+/datum/tgui_bug_report_form/proc/send_request(payload_body, client/user)
+	// for any future changes see https://docs.github.com/en/rest/issues/issues
+	var/repo_name = CONFIG_GET(string/repo_name)
+	var/org = CONFIG_GET(string/org)
+	var/token = CONFIG_GET(string/github_app_api)
+
+	if(!token || !org || !repo_name)
+		tgui_alert(user, "The configuration is not set for the external api", "Issue not reported!")
```
Comment (harryob) [inline code/datums/bug_report.dm:87]:
```suggestion
		tgui_alert(user, "The configuration is not set for the external API.", "Issue not reported!")
```

File: code/datums/bug_report.dm [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+		return
+
+	var/url = "https://api.github.com/repos/[org]/[repo_name]/issues"
+	var/list/headers = list()
+
+	headers["Authorization"] = "Bearer [token]"
+	headers["Content-Type"] = "text/markdown; charset=utf-8"
+	headers["Accept"] = "application/vnd.github+json"
+
+	var/datum/http_request/request = new
```
Comment (harryob) [inline code/datums/bug_report.dm:99]:
```suggestion
	var/datum/http_request/request = new()
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/BugReportForm.tsx (hunk 6 above)
Comment (vero5123) [reply tgui/packages/tgui/interfaces/BugReportForm.tsx:47]: Nowhere, Its use was factored out earlier when I switched to useState instead; I forgot to remove it altogether.

File: code/datums/bug_report.dm (hunk 7 above)
Comment (vero5123) [reply code/datums/bug_report.dm]: done

**Comment by vero5123** [comment]
I'm happy where it's at code wise and have done lots of testing, afik there ~~shouldn't~~ be any runtimes.

File: code/datums/bug_report.dm [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	var/list/bug_report_data = null
+
+	// client of user who created the initial report, immutable, set on init.
+	var/client/initial_user = null
+
+	// client of the admin who is accessing the report, we don't want multiple admins unknowingly making changes at the same time.
+	var/client/admin_user = null
+
+	// value to determine if the bug report is submitted and awaiting admin approval, used for state purposes in tgui.
+	var/awaiting_admin_approval = FALSE
```
Comment (Drulikar) [inline code/datums/bug_report.dm:15]:
```suggestion
	/// contains all the body text for the bug report.
	var/list/bug_report_data = null

	/// client of user who created the initial report, immutable, set on init.
	var/client/initial_user = null

	/// client of the admin who is accessing the report, we don't want multiple admins unknowingly making changes at the same time.
	var/client/admin_user = null

	/// value to determine if the bug report is submitted and awaiting admin approval, used for state purposes in tgui.
	var/awaiting_admin_approval = FALSE
```

File: code/datums/bug_report.dm [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+	// for garbage collection purposes.
+	var/selected_confirm = FALSE
+
+/datum/tgui_bug_report_form/New(mob/user)
+	if(!user.client) // nope.
+		return INITIALIZE_HINT_QDEL
```
Comment (Drulikar) [inline code/datums/bug_report.dm:22]: You can only do this in Initialize. As I mentioned its fine to set the initial_user here in New but then in Initialize check if initial_user is valid. if its not, then return `INITIALIZE_HINT_QDEL`

File: code/datums/bug_report.dm [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+		return
+
+	if(tgui_alert(user, "This will open the GitHub in your browser. Are you sure?", "Confirm", list("Yes", "No")) == "Yes")
+		user << link(url)
+
+/datum/tgui_bug_report_form/ui_state()
+	return GLOB.always_state
+
+/datum/tgui_bug_report_form/tgui_interact(mob/user, datum/tgui/ui)
+
```
Comment (Drulikar) [inline code/datums/bug_report.dm:39]:
```suggestion
```

**Review by Drulikar — CHANGES_REQUESTED** [review]

File: code/datums/bug_report.dm (hunk 11 above)
Comment (vero5123) [reply code/datums/bug_report.dm:22]: INITIALIZE_HINT_QDEL is for atoms, it's part of the atom subsystem.

**Review by harryob — APPROVED** [review]

**Pull request merged**